### PR TITLE
Add BayesianScoreQuery and LogOddsFusionQuery for probabilistic hybrid search

### PR DESCRIPTION
**Pull request by jaepil**
## Summary

Add `BayesianScoreQuery` and `LogOddsFusionQuery` to Lucene, enabling probabilistic hybrid search that combines BM25 text retrieval with KNN vector search (or any other probability-producing scorers) through a principled Bayesian framework.

Reference implementation: [cognica-io/bayesian-bm25](https://github.com/cognica-io/bayesian-bm25)

**This PR adds 5 new files and modifies only CHANGES.txt.**

## Motivation

Standard BM25 scores are unbounded and incomparable across queries. KNN cosine similarities live in [0, 1]. Combining these heterogeneous signals typically requires ad-hoc normalization (e.g., RRF, linear interpolation). This PR provides a theoretically grounded alternative:

1. **BayesianScoreQuery** wraps any query and applies sigmoid calibration to transform scores into probabilities in (0, 1)
2. **LogOddsFusionQuery** combines multiple probability signals via log-odds, resolving the shrinkage problem of naive probabilistic AND

Based on "Bayesian BM25" and "From Bayesian Inference to Neural Computation".

### BayesianScoreQuery

A Query wrapper that applies sigmoid calibration to the inner query's score:

```
P = sigmoid(alpha * (score - beta))
```

The inner query (typically a multi-term BooleanQuery with BM25Similarity) produces a raw score via its standard IDF-weighted term combination. BayesianScoreQuery maps this query-level score to a probability in (0, 1). Since sigmoid is monotone, the original BM25 ranking is preserved exactly.

### LogOddsFusionQuery

Extends `Query`, following `DisjunctionMaxQuery` pattern. Inner `LogOddsFusionScorer` extends `DisjunctionScorer`.

```
for each matching sub-scorer:
    gatedLogit = softplus(logit(clamp(subScore)))
logitSum    = sum of gatedLogit values
meanLogit   = logitSum / n          (n = total clauses, not just matching)
scaledLogit = meanLogit * n^alpha   (default alpha = 0.5, sqrt(n) scaling)
score       = sigmoid(scaledLogit)
```

Non-matching sub-scorers contribute 0 (neutral evidence). The `alpha` parameter controls confidence scaling: `alpha=0.5` (default) implements the sqrt(n) scaling law where confidence grows as the square root of agreeing signals.

### Query-level calibration (not Similarity-level)

An earlier iteration implemented the Bayesian transform inside a custom `Similarity`, applying it per-term. This broke BM25's IDF-weighted term combination: posterior values in (0,1) lost the dynamic range that IDF encodes in raw BM25 scores, causing significant NDCG degradation on BEIR benchmarks. The paper's design operates at the query level -- BM25 first combines terms via its standard sum, then a single sigmoid maps the result to a probability. `BayesianScoreQuery` implements this correctly as a Query wrapper.

### Softplus gating

The reference `log_odds_conjunction()` applies no gating by default (`gating="none"`), with ReLU/Swish/GELU as options. In a Lucene disjunction, a non-matching sub-scorer contributes `logit(0.5) = 0` (neutral). Without gating, a matching sub-scorer with probability < 0.5 contributes a negative logit -- meaning a document is penalized for *containing* a query term. This conflates "absence of evidence" with "evidence of absence."

- **ReLU** (`max(0, logit)`) fixes this but collapses all weak matches to exactly 0, losing ordering among them
- **Swish/GELU** suppress negative logits but still produce negative values
- **Softplus** (`log(1 + exp(logit))`) is always positive, preserving the ordering among weak matches while guaranteeing that any match contributes more than no match

Softplus is hardcoded rather than configurable because the distinction between "no match" and "weak match" is structural in Lucene's disjunction model, not a tuning parameter.

### Default codec for KNN tests

Vector tests use `TestUtil.getDefaultCodec()` to avoid randomized codecs that may produce unreliable HNSW graphs, matching the pattern in Lucene's `BaseKnnVectorQueryTestCase`.

## Integration with BooleanQuery

`LogOddsFusionQuery` produces scores in (0, 1) and follows standard Lucene `Query` semantics, so it composes naturally with `BooleanQuery`:

| Pattern | Lucene Expression | Semantics |
|---------|-------------------|-----------|
| Hybrid AND filter | `hybrid MUST, text FILTER` | Only text-matching docs scored by hybrid |
| Hybrid OR hybrid | `hybrid1 SHOULD, hybrid2 SHOULD` | Match either hybrid, both boost score |
| Hybrid NOT vector | `text MUST, knn MUST_NOT` | Exclude semantically similar docs from text results |
| Multi-signal fusion | `LogOddsFusion(BayesianScore(text), knn1, knn2)` | Three-way text + vector + vector combination |

## New Files

| File | Description |
|------|-------------|
| `BayesianScoreQuery.java` | Query wrapper applying sigmoid calibration to produce (0,1) probabilities |
| `LogOddsFusionQuery.java` | Query combining probability scores via log-odds conjunction |
| `LogOddsFusionScorer.java` | DisjunctionScorer with softplus-gated log-odds scoring |
| `TestBayesianScoreQuery.java` | Tests for BayesianScoreQuery (12 tests) |
| `TestLogOddsFusionQuery.java` | Tests for LogOddsFusion + hybrid search (22 tests) |

### BayesianScoreQuery (12 tests)

- Score output in (0,1) probability range
- Single-term and multi-term ranking preservation (sigmoid monotonicity)
- Sigmoid numerical stability (large positive/negative inputs, zero)
- Parameter validation, rewrite, equals/hashCode, toString
- Explanation correctness, QueryUtils.check, CheckHits.checkTopScores
- Different alpha/beta preserve ranking order

### LogOddsFusionQuery (22 tests)

**Core query behavior:**
- Basic fusion, single/empty clause rewrite, alpha validation
- equals/hashCode, toString, QueryUtils.check
- Explanation correctness, WAND max score correctness
- Three-way combination

**Multi-field text hybrid:**
- Multi-field hybrid (title + body via BayesianScoreQuery)
- More signals increase score (softplus anti-shrinkage property)

**Vector + text hybrid:**
- Text + vector combination: dual-signal doc ranks above single-signal
- Both text-only and vector-only signals contribute to results
- Multiple vector fields combined via LogOdds

**Boolean logic combinations:**
- MUST (AND): text filter + hybrid scoring
- SHOULD (OR): two independent hybrids combined
- MUST_NOT (NOT): exclude vector-similar docs from text results
- Vector AND vector: MUST conjunction of two KNN queries
- Vector NOT: KNN MUST_NOT exclusion

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/search/BayesianScoreQuery.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+ * calibration: {@code P = sigmoid(alpha * (score - beta))}.
+ *
+ * <p>This implements the query-level Bayesian transform from "Bayesian BM25": the inner query
+ * (typically a multi-term BooleanQuery with BM25Similarity) produces a raw score, and this wrapper
+ * maps it to a probability in (0, 1) suitable for combination with other probability signals via
+ * {@link LogOddsConjunctionQuery}.
+ *
+ * <p>The alpha parameter controls the sigmoid steepness (score sensitivity), and beta controls the
+ * midpoint (decision boundary). These can be set manually or estimated from the score distribution.
+ */
```
Comment (benwtrent) [inline lucene/core/src/java/org/apache/lucene/search/BayesianScoreQuery.java:36]:
```suggestion
 * @lucene.experimental
 */
```

File: lucene/core/src/java/org/apache/lucene/search/LogOddsConjunctionQuery.java [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+ *   <li>Converting each sub-score to log-odds: logit(p) = log(p / (1 - p))
+ *   <li>Computing the mean log-odds across all clauses (non-matching contribute 0 = neutral)
+ *   <li>Applying multiplicative confidence scaling: meanLogit * n^alpha
+ *   <li>Converting back to probability via sigmoid
+ * </ol>
+ *
+ * <p>The alpha parameter controls the confidence scaling exponent. The default alpha=0.5 implements
+ * the sqrt(n) scaling law from "From Bayesian Inference to Neural Computation".
+ *
+ * @see LogOddsConjunctionScorer
```
Comment (benwtrent) [inline lucene/core/src/java/org/apache/lucene/search/LogOddsConjunctionQuery.java]:
```suggestion
 * @see LogOddsConjunctionScorer
 * @lucene.experimental
```

File: lucene/core/src/java/org/apache/lucene/search/LogOddsFusionScorer.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+ *   score       = sigmoid(scaledLogit)
+ * </pre>
+ *
+ * <p>Softplus gating ({@code log(1 + exp(x))}) is applied to logit values before aggregation. This
+ * distinguishes "absence of evidence" (non-matching sub-scorer, contributes 0) from "evidence of
+ * absence" (matching sub-scorer with weak probability, contributes a small positive value). A
+ * matching sub-scorer always contributes more than a non-matching one, preserving the ordering
+ * among weak matches while ensuring that no match is ever penalized.
+ *
+ * <p>Non-matching sub-scorers contribute logit(0.5) = 0 (neutral evidence).
```
Comment (benwtrent) [inline lucene/core/src/java/org/apache/lucene/search/LogOddsFusionScorer.java:42]:
```suggestion
 * <p>Non-matching sub-scorers contribute logit(0.5) = 0 (neutral evidence).
 * @lucene.experimental
```

File: lucene/core/src/java/org/apache/lucene/search/LogOddsConjunctionScorer.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    this.scalingFactor = (float) Math.pow(totalClauses, alpha);
+    if (scoreMode == ScoreMode.TOP_SCORES) {
+      this.disjunctionBlockPropagator = new DisjunctionScoreBlockBoundaryPropagator(subScorers);
+    } else {
+      this.disjunctionBlockPropagator = null;
+    }
+  }
+
+  static float clampProbability(float p) {
+    return Math.max(CLAMP_MIN, Math.min(CLAMP_MAX, p));
```
Comment (benwtrent) [inline lucene/core/src/java/org/apache/lucene/search/LogOddsConjunctionScorer.java]: We can use `Math.clamp`

**Review by benwtrent — COMMENTED** [review]
Thank you for the contribution! I have been interested in y'alls bayesian bm25 :)

Some minor quibbles so far (need to review more). 

Maybe flag as 10.5 and move the changes entry to there?

Also, these seem like Disjunctions not Conjunctions? Maybe I am reading it wrong?

File: lucene/core/src/java/org/apache/lucene/search/LogOddsConjunctionScorer.java (hunk 4 above)
Comment (jaepil) [reply lucene/core/src/java/org/apache/lucene/search/LogOddsConjunctionScorer.java]: Done. Thank you!

**Comment by jaepil** [comment]
@benwtrent Thank you for the review!

Good point -- the naming reflects two different layers:

- Document matching: yes, this is a disjunction. It extends DisjunctionScorer and any matching sub-query contributes to the result.
- Scoring model: "conjunction" refers to the probabilistic operation -- combining evidence via log-odds addition, which is the Bayesian mechanism for computing P(A AND B) under conditional independence.

So the name describes what the scoring computes (probabilistic conjunction of evidence), not how documents are matched (which is indeed a disjunction).

That said, I can see how this is confusing in Lucene's context where Conjunction/Disjunction have established meanings tied to document matching. Happy to rename if that helps -- something like `LogOddsFusionQuery`/`LogOddsFusionScorer` would avoid the ambiguity while staying descriptive of the hybrid search use case. What do you think?

**Comment by benwtrent** [comment]
@jaepil 

Ah, thank you for the explanation. Yes, OK, its conjunction in terms of "boosting" (obvious simplification ;)) the scores.

`LogOddsFusionQuery` and same for the scorer works just fine. 

Also, the reason I suggested labeling as "experimental" as that allows the API or behavior to change if we deem it necessary after more testing within Lucene.

**Comment by jaepil** [comment]
@benwtrent Thanks for the review! Renamed LogOddsConjunctionQuery/Scorer to LogOddsFusionQuery/Scorer.

**Review by benwtrent — APPROVED** [review]
I like it, it looks pretty good and would be good for folks to kick the tires and see how it performs on their data!

**Comment by jaepil** [comment]
Thank you! I'm looking forward to seeing how this improves hybrid score fusion, especially in multimodal setups.

**Comment by benwtrent** [comment]
Thank you for your contribution @jaepil I will merge and backport after letting some more time for others if they want to review.

**Comment by jaepil** [comment]
@benwtrent Thank you for letting me know about the process!
